sup3r/pipeline/forward_pass.py

```diff
@@ -1128,13 +1113,210 @@ def __init__(self, strategy, chunk_index=0, node_index=0):
+            Resolution of exogenous data. e.g. {'temporal': 15min, 'spatial':
+            '1km'}
+
+        Returns
+        -------
+        updated_step_dict : dict
+            Same as input dictionary with s_agg_factor, t_agg_factor,
+            s_enhance, t_enhance added
+        """
+        model_step = step_dict['model']
+        s_enhance = None
+        t_enhance = None
+        s_agg_factor = None
+        t_agg_factor = None
+        models = getattr(self.model, 'models', [self.model])
+        msg = (f'Model index from exo_kwargs ({model_step} exceeds number '
+               f'of model steps ({len(models)})')
+        assert len(models) > model_step, msg
+        model = models[model_step]
+        input_res = model.input_resolution
+        output_res = model.output_resolution
+        combine_type = step_dict.get('combine_type', None)
+
+        if combine_type.lower() == 'input':
+            if model_step == 0:
+                s_enhance = 1
+                t_enhance = 1
+            else:
+                s_enhance = np.product(
+                    self.strategy.s_enhancements[:model_step])
+                t_enhance = np.product(
+                    self.strategy.t_enhancements[:model_step])
+            s_agg_factor, t_agg_factor = self.get_agg_factors(
+                input_res, exo_resolution)
+            resolution = input_res
+
+        elif combine_type.lower() in ('output', 'layer'):
+            s_enhance = np.product(
+                self.strategy.s_enhancements[:model_step + 1])
+            t_enhance = np.product(
+                self.strategy.t_enhancements[:model_step + 1])
+            s_agg_factor, t_agg_factor = self.get_agg_factors(
+                output_res, exo_resolution)
+            resolution = output_res
+
+        else:
+            msg = ('Received exo_kwargs entry without valid combine_type '
+                   '(input/layer/output)')
+            raise OSError(msg)
+
+        updated_dict = step_dict.copy()
+        updated_dict.update({'s_enhance': s_enhance, 't_enhance': t_enhance,
+                             's_agg_factor': s_agg_factor,
+                             't_agg_factor': t_agg_factor,
+                             'resolution': resolution})
+        return updated_dict
+
+    def update_exo_extract_kwargs(self, exo_kwargs):
+        """Compute agg and enhancement factors for all model steps for all
+        features.
+
+        Parameters
+        ----------
+        exo_kwargs: dict
+            Full exo_kwargs dictionary with all feature entries.
+            e.g. {'topography': {'exo_resolution': {'spatial': '1km',
+            'temporal': None}, 'steps': [{'model': 0, 'combine_type': 'input'},
+            {'model': 0, 'combine_type': 'layer'}]}}
+
+        Returns
+        -------
+        updated_dict : dict
+            Same as input dictionary with s_agg_factor, t_agg_factor,
+            s_enhance, t_enhance added to each step entry for all features
```

Comment:
FWIW i would still prefer to input the agg factors explicitly so we can tweak them as a sensitivity and have less code logic.